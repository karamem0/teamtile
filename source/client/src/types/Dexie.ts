//
// Copyright (c) 2023 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export interface ArrayEntity<T> {
  id: string,
  expired: number,
  values: T[]
}

export interface ValueEntity<T> {
  id: string,
  expired: number,
  value: T
}
