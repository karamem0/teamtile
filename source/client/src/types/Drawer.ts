//
// Copyright (c) 2021-2025 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export type DrawerType =
  | 'channel'
  | 'member'
  | 'tag';
