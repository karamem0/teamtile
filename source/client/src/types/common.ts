//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export interface KeyValue<K, V> {
  key: K,
  value: V
}
