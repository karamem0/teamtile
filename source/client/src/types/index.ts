//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

export interface Team {
  id?: string;
  internalId?: string;
  name?: string;
  description?: string;
  visibility?: string;
  url?: string;
  icon?: string;
}

export interface TeamChannels {
  count?: number;
  nextLink?: string;
  values?: TeamChannel[];
}

export interface TeamChannel {
  id?: string;
  name?: string;
  url?: string;
}

export interface TeamMembers {
  count?: number;
  nextLink?: string;
  values?: TeamMember[];
}

export interface TeamMember {
  id?: string;
  name?: string;
  email?: string;
}

export interface TeamDrive {
  id?: string;
  url?: string;
}
