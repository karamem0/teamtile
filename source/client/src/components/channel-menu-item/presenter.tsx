//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { ContextMenuIcon } from '@fluentui/react-icons-mdl2';
import {
  List,
  Text
} from '@fluentui/react-northstar';

import { css } from '@emotion/react';

import { EventHandler } from '../../types/common';
import { Channel } from '../../types/entity';
import { CardMenuItem } from '../card-menu-item';
import { CardPopup } from '../card-popup';
import { MembershipIcon } from '../membership-icon';

interface ChannelMenuItemProps {
  channels: Channel[],
  onClick?: EventHandler<Channel> | undefined
}

export default React.memo(function ChannelMenuItem ({
  channels,
  onClick
}: ChannelMenuItemProps): React.ReactElement | null {

  return (
    <CardPopup<Channel>
      items={channels}
      predicate={(filter, item) => {
        if (item.displayName) {
          if (item.displayName.search(new RegExp(filter, 'i')) >= 0) {
            return true;
          }
        }
        return false;
      }}
      renderer={(items) => (
        <List
          items={
            items.map((item) => ({
              key: item.id,
              header: (
                <Text
                  css={css`
                    display: grid;
                    grid-template-columns: auto auto;
                    gap: 0.25rem;
                    align-items: center;
                    justify-content: left;
                    margin: 0 -0.5rem;
                  `}
                  role="button"
                  onClick={(event: React.SyntheticEvent) => onClick && onClick(event, item)}>
                  <Text truncated>
                    {item.displayName}
                  </Text>
                  <MembershipIcon membership={item.membershipType} />
                </Text>
              )
            }))
          }
          navigable />
      )}
      trigger={
        <CardMenuItem
          content={channels.length}
          icon={
            <ContextMenuIcon
              css={css`
                width: 1rem;
                height: 1rem;
              `} />
          } />
      } />
  );

});
