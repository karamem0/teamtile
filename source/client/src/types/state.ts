//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import { KeyValue } from './common';
import {
  Channel,
  Drive,
  Icon,
  Member,
  Team
} from './entity';

export type ItemKey = string;

export type ItemValue = (
  Team &
  Icon &
  {
    channels: Channel[] | undefined,
    members: (Member & Icon)[] | undefined,
    drive: Drive | undefined
  }
);

export type Item = (
  KeyValue<ItemKey, ItemValue | null> &
  {
    visible: boolean | undefined
  }
);

export enum Loading {
  none = 0,
  keys = 1,
  values = 2,
  done = 255,
}

export interface State {
  loading: Loading,
  itemFilter: string | null | undefined,
  items: Item[]
}
