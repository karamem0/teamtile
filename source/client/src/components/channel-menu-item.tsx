//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { ContextMenuIcon } from '@fluentui/react-icons-mdl2';
import {
  List,
  Text
} from '@fluentui/react-northstar';

import { app } from '@microsoft/teams-js';

import { css } from '@emotion/react';

import { KeyValue } from '../types/common';
import { Channel } from '../types/entity';
import { ItemKey, ItemValue } from '../types/state';

import { CardMenuItem } from './card-menu-item';
import { CardPopup } from './card-popup';
import { MembershipIcon } from './membership-icon';

export interface ChannelMenuItemProps {
  item: KeyValue<ItemKey, ItemValue>
}

export const ChannelMenuItem = ({ item }: ChannelMenuItemProps): React.ReactElement | null => {

  if (!item.value.channels) {
    return null;
  }

  return (
    <ChannelMenuItemPresenterMemo
      channels={item.value.channels}
      onClick={(value) => app.openLink(value)} />
  );

};

interface ChannelMenuItemPresenterProps {
  channels: Channel[],
  onClick: (value: string) => void
}

const ChannelMenuItemPresenter = ({
  channels,
  onClick
}: ChannelMenuItemPresenterProps): React.ReactElement | null => {

  return (
    <CardPopup
      predicate={(filter, value) => {
        if (value.displayName) {
          if (value.displayName.search(new RegExp(filter, 'i')) >= 0) {
            return true;
          }
        }
        return false;
      }}
      renderer={(values) => (
        <List
          items={
            values.map((value) => ({
              key: value.id,
              header: (
                <Text
                  css={css`
                    display: grid;
                    grid-template-columns: auto auto;
                    gap: 0.25rem;
                    align-items: center;
                    justify-content: left;
                    margin: 0 -0.5rem 0 -0.5rem;
                  `}
                  role="button"
                  onClick={() => value.webUrl && onClick(value.webUrl)}>
                  <Text truncated>
                    {value.displayName}
                  </Text>
                  <MembershipIcon membership={value.membershipType} />
                </Text>
              )
            }))
          }
          navigable />
      )}
      trigger={
        <CardMenuItem
          content={channels.length}
          icon={<ContextMenuIcon />} />
      }
      values={channels} />
  );

};

const ChannelMenuItemPresenterMemo = React.memo(ChannelMenuItemPresenter);
