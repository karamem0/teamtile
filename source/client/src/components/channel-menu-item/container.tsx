//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { app } from '@microsoft/teams-js';

import { Channel } from '../../types/entity';
import { ItemKey, ItemValue } from '../../types/state';

import Presenter from './presenter';

interface ChannelMenuItemProps {
  itemKey: ItemKey,
  itemValue: ItemValue
}

export default function ChannelMenuItem ({
  itemValue
}: ChannelMenuItemProps): React.ReactElement | null {

  const handleClick = React.useCallback((_, data: Channel | undefined) => {
    if (!data?.webUrl) {
      return;
    }
    app.openLink(data.webUrl);
  }, []);

  if (!itemValue.channels) {
    return null;
  }

  return (
    <Presenter
      channels={itemValue.channels}
      onClick={handleClick} />
  );

}
