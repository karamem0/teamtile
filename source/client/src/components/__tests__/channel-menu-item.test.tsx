//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { render, screen } from '@testing-library/react';

import { ItemValue } from '../../types/state';
import { ChannelMenuItem } from '../channel-menu-item';

beforeEach(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
});

describe('ChannelMenuItem', () => {

  it('create shapshot', async () => {
    const json = await import('./__jsons__/channel-menu-item.json');
    const params = {
      item: {
        key: '02bd9fd6-8f93-4758-87c3-1fb73740a315',
        value: json.default as ItemValue
      }
    };
    render(<ChannelMenuItem {...params} />);
    expect(screen.queryAllByText(/^.*$/)[0]).toMatchSnapshot();
  });

});
