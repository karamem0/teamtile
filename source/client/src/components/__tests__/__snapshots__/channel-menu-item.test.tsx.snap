// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelMenuItem create shapshot 1`] = `
.emotion-0 {
  display: grid;
  grid-template-columns: auto auto;
  grid-template-rows: auto;
  grid-gap: 0.25rem;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  cursor: pointer;
}

.emotion-1 {
  line-height: 1rem;
  width: 1rem;
  height: 1rem;
}

.emotion-2 {
  line-height: 0.75rem;
}

<body>
  <div>
    <span
      class="ui-text emotion-0"
      data-uses-unhanded-props="true"
      role="button"
    >
      <span
        class="emotion-1"
      >
        <span
          aria-hidden="true"
          class="root_3aeb045a"
        >
          <svg
            class="svg_3aeb045a"
            focusable="false"
            viewBox="0 0 2048 2048"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1792 128v1664H256V128h1536zm-128 128H384v1408h1280V256zm-256 768H640V896h768v128zm0 384H640v-128h768v128zm0-768H640V512h768v128z"
            />
          </svg>
        </span>
      </span>
      <span
        class="ui-text emotion-2"
        data-uses-unhanded-props="true"
      >
        3
      </span>
    </span>
  </div>
</body>
`;
